File: process.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"math"
	"os"
	"path/filepath"
	"strconv"
)

const C = 299792458.0 // speed of light (m/s)

// --- Helper Functions ---

func norm2(x, y float64) float64 {
	return math.Sqrt(x*x + y*y)
}

func lstsq3x3(J [3][3]float64, r [3]float64) (dx, dy, dt float64, ok bool) {
	// Solve J * delta = -r using Gaussian elimination
	A := J
	b := [3]float64{-r[0], -r[1], -r[2]}

	for i := 0; i < 3; i++ {
		// pivot
		maxRow := i
		for k := i + 1; k < 3; k++ {
			if math.Abs(A[k][i]) > math.Abs(A[maxRow][i]) {
				maxRow = k
			}
		}
		A[i], A[maxRow] = A[maxRow], A[i]
		b[i], b[maxRow] = b[maxRow], b[i]

		if math.Abs(A[i][i]) < 1e-15 {
			return 0, 0, 0, false
		}

		// eliminate
		for k := i + 1; k < 3; k++ {
			f := A[k][i] / A[i][i]
			for j := i; j < 3; j++ {
				A[k][j] -= f * A[i][j]
			}
			b[k] -= f * b[i]
		}
	}

	// back substitution
	x := [3]float64{}
	for i := 2; i >= 0; i-- {
		sum := b[i]
		for j := i + 1; j < 3; j++ {
			sum -= A[i][j] * x[j]
		}
		x[i] = sum / A[i][i]
	}
	return x[0], x[1], x[2], true
}

// --- Gauss–Newton TDOA Solver (multi-start) ---

type Solution struct {
	X, Y, T0   float64
	Converged  bool
	Iterations int
	ResNorm    float64
}

func GaussNewtonTDOAMultiStart(gwCoords [3][2]float64, t_ns [3]float64, starts [][2]float64, maxIter int, tol float64) Solution {
	tMeas := [3]float64{}
	for i := 0; i < 3; i++ {
		tMeas[i] = t_ns[i] * 1e-9
	}

	best := Solution{ResNorm: math.Inf(1)}

	for _, start := range starts {
		x, y := start[0], start[1]
		centroid := [2]float64{x, y}

		minD := math.Inf(1)
		for _, gw := range gwCoords {
			d := norm2(gw[0]-centroid[0], gw[1]-centroid[1])
			if d < minD {
				minD = d
			}
		}
		t0 := min(tMeas[0], min(tMeas[1], tMeas[2])) - minD/C

		xx, yy, tt0 := x, y, t0

		for k := 0; k < maxIter; k++ {
			var ranges [3]float64
			for i := 0; i < 3; i++ {
				ranges[i] = norm2(xx-gwCoords[i][0], yy-gwCoords[i][1])
			}

			var pred, residuals [3]float64
			for i := 0; i < 3; i++ {
				pred[i] = tt0 + ranges[i]/C
				residuals[i] = pred[i] - tMeas[i]
			}

			var J [3][3]float64
			for i := 0; i < 3; i++ {
				ri := ranges[i]
				if ri != 0 {
					J[i][0] = (xx - gwCoords[i][0]) / (ri * C)
					J[i][1] = (yy - gwCoords[i][1]) / (ri * C)
				}
				J[i][2] = 1.0
			}

			dx, dy, dt, ok := lstsq3x3(J, residuals)
			if !ok {
				break
			}

			xx += dx
			yy += dy
			tt0 += dt

			if math.Sqrt(dx*dx+dy*dy+dt*dt) < tol {
				resNorm := 0.0
				for _, r := range residuals {
					resNorm += r * r
				}
				resNorm = math.Sqrt(resNorm)
				if resNorm < best.ResNorm {
					best = Solution{xx, yy, tt0, true, k + 1, resNorm}
				}
				break
			}
		}
	}
	return best
}

// --- Dataset Processor ---

func ProcessTDOADataset(csvPath string) []map[string]interface{} {
	file, err := os.Open(csvPath)
	if err != nil {
		log.Fatalf("failed to open CSV: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("failed to read CSV: %v", err)
	}

	header := rows[0]
	data := rows[1:]

	col := func(name string) int {
		for i, h := range header {
			if h == name {
				return i
			}
		}
		log.Fatalf("missing column: %s", name)
		return -1
	}

	// Extract gateway coordinates
	gwCoords := [3][2]float64{}
	for i := 0; i < 3; i++ {
		x, _ := strconv.ParseFloat(data[0][col(fmt.Sprintf("gw%d_x_m", i+1))], 64)
		y, _ := strconv.ParseFloat(data[0][col(fmt.Sprintf("gw%d_y_m", i+1))], 64)
		gwCoords[i] = [2]float64{x, y}
	}

	results := []map[string]interface{}{}

	for i, row := range data {
		var t_ns [3]float64
		for j := 0; j < 3; j++ {
			t_ns[j], _ = strconv.ParseFloat(row[col(fmt.Sprintf("t_gw%d_ns", j+1))], 64)
		}

		starts := [][2]float64{
			{(gwCoords[0][0] + gwCoords[1][0] + gwCoords[2][0]) / 3,
				(gwCoords[0][1] + gwCoords[1][1] + gwCoords[2][1]) / 3},
			{gwCoords[0][0], gwCoords[0][1]},
			{gwCoords[1][0], gwCoords[1][1]},
			{gwCoords[2][0], gwCoords[2][1]},
		}

		sol := GaussNewtonTDOAMultiStart(gwCoords, t_ns, starts, 100, 1e-9)

		res := map[string]interface{}{
			"sample_id":         i + 1,
			"recov_x_m":         sol.X,
			"recov_y_m":         sol.Y,
			"solver_converged":  sol.Converged,
			"solver_iters":      sol.Iterations,
			"solver_resnorm_s":  sol.ResNorm,
		}
		results = append(results, res)
	}

	outPath := "tdoa_results.csv"
	f, err := os.Create(outPath)
	if err != nil {
		log.Fatalf("failed to create output CSV: %v", err)
	}
	defer f.Close()

	writer := csv.NewWriter(f)
	defer writer.Flush()

	writer.Write([]string{"sample_id", "recov_x_m", "recov_y_m", "solver_converged", "solver_iters", "solver_resnorm_s"})
	for _, r := range results {
		writer.Write([]string{
			fmt.Sprintf("%v", r["sample_id"]),
			fmt.Sprintf("%v", r["recov_x_m"]),
			fmt.Sprintf("%v", r["recov_y_m"]),
			fmt.Sprintf("%v", r["solver_converged"]),
			fmt.Sprintf("%v", r["solver_iters"]),
			fmt.Sprintf("%v", r["solver_resnorm_s"]),
		})
	}

	abs, _ := filepath.Abs(outPath)
	fmt.Printf("✅ Results saved to: %s\n", abs)
	return results
}

// --- Main Entry Point ---

func main() {
	inputCSV := "lorawan_data.csv"
	ProcessTDOADataset(inputCSV)
}

// --- Utilities ---

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

File: process.py
import numpy as np
import pandas as pd
import math
from pathlib import Path

# --- Constants ---
C = 299_792_458.0  # speed of light (m/s)

# --- Gauss–Newton TDOA Solver (multi-start) ---
def gauss_newton_tdoa_multistart(gw_coords, t_ns, starts, max_iter=100, tol=1e-9):
    """Estimate (x, y, t0) of node from gateway nanosecond timestamps."""
    t_meas = np.array(t_ns, dtype=float) * 1e-9
    best_sol = (None, None, None, False, 0, np.inf)

    for start in starts:
        if len(start) == 3:
            x, y, t0 = start
        else:
            x, y = start
            centroid = np.array([x, y])
            d_to_centroid = np.linalg.norm(gw_coords - centroid, axis=1).min()
            t0 = t_meas.min() - d_to_centroid / C

        xx, yy, tt0 = x, y, t0

        for k in range(max_iter):
            ranges = np.sqrt((xx - gw_coords[:, 0])**2 + (yy - gw_coords[:, 1])**2)
            pred = tt0 + ranges / C
            residuals = pred - t_meas

            J = np.zeros((3, 3))
            for i in range(3):
                ri = ranges[i]
                if ri == 0:
                    J[i, 0] = J[i, 1] = 0.0
                else:
                    J[i, 0] = (xx - gw_coords[i, 0]) / (ri * C)
                    J[i, 1] = (yy - gw_coords[i, 1]) / (ri * C)
                J[i, 2] = 1.0

            try:
                delta, *_ = np.linalg.lstsq(J, -residuals, rcond=None)
            except np.linalg.LinAlgError:
                break

            xx += float(delta[0])
            yy += float(delta[1])
            tt0 += float(delta[2])

            if np.linalg.norm(delta) < tol:
                resnorm = np.linalg.norm(residuals)
                if resnorm < best_sol[5]:
                    best_sol = (xx, yy, tt0, True, k + 1, resnorm)
                break
        else:
            ranges = np.sqrt((xx - gw_coords[:, 0])**2 + (yy - gw_coords[:, 1])**2)
            pred = tt0 + ranges / C
            residuals = pred - t_meas
            resnorm = np.linalg.norm(residuals)
            if resnorm < best_sol[5]:
                best_sol = (xx, yy, tt0, False, max_iter, resnorm)
    return best_sol

# --- Dataset Processor ---
def process_tdoa_dataset(csv_path):
    """
    Reads dataset with gateway timestamp columns (t_gw1_ns, t_gw2_ns, t_gw3_ns)
    and solves for the node location using Gauss–Newton TDOA method.
    """
    df = pd.read_csv(csv_path)

    # Extract gateway coordinates (assumed constant for all samples)
    gw_coords = np.array([
        [df["gw1_x_m"].iloc[0], df["gw1_y_m"].iloc[0]],
        [df["gw2_x_m"].iloc[0], df["gw2_y_m"].iloc[0]],
        [df["gw3_x_m"].iloc[0], df["gw3_y_m"].iloc[0]],
    ])

    results = []
    for i, row in df.iterrows():
        t_ns = [row["t_gw1_ns"], row["t_gw2_ns"], row["t_gw3_ns"]]
        starts = [
            (gw_coords[:, 0].mean(), gw_coords[:, 1].mean()),
            (gw_coords[0, 0], gw_coords[0, 1]),
            (gw_coords[1, 0], gw_coords[1, 1]),
            (gw_coords[2, 0], gw_coords[2, 1]),
        ]

        sol_x, sol_y, sol_t0, converged, iters, resnorm = gauss_newton_tdoa_multistart(
            gw_coords, t_ns, starts
        )

        if sol_x is not None and converged:
            results.append({
                "sample_id": row.get("sample_id", i + 1),
                "recov_x_m": sol_x,
                "recov_y_m": sol_y,
                "solver_converged": converged,
                "solver_iters": iters,
                "solver_resnorm_s": resnorm,
            })
        else:
            results.append({
                "sample_id": row.get("sample_id", i + 1),
                "recov_x_m": np.nan,
                "recov_y_m": np.nan,
                "solver_converged": False,
                "solver_iters": iters,
                "solver_resnorm_s": resnorm,
            })

    out_df = pd.DataFrame(results)
    out_path = Path("tdoa_results.csv")
    out_df.to_csv(out_path, index=False)
    print(f"✅ Results saved to: {out_path.resolve()}")
    print(out_df.head())
    return out_df

# --- Main Entry Point ---
if __name__ == "__main__":
    input_csv = "lorawan_data.csv"
    df_results = process_tdoa_dataset(input_csv)
